### ISSUE TITLE
Unable to upload file using Element Send Keys

### ISSUE THREAD

**Issue by MakarandDhuri** [mozilla/geckodriver#168]
OS: ubuntu 14.04
Browser:firefox 48.0
selenium-server-standalone.jar:selenium-server-standalone-3.0.0-beta2.jar

Scenario: 

I am using FirefoxDriver with geckodriver, same code working fine with ChromeDriver.

code: 

```
public static void main(String[] args) {
        System.setProperty("webdriver.gecko.driver", "src/main/resources/geckodriver");
        WebDriver driver = new FirefoxDriver();
        driver.get("https://how-old.net/");
        ((JavascriptExecutor) driver).executeScript(
                "arguments[0].style.visibility = 'visible'; arguments[0].style.height = '1px'; arguments[0].style.width = '1px'; arguments[0].style.opacity = 1",
                driver.findElement(By.id("uploadBtn")));

        driver.findElement(By.id("uploadBtn")).sendKeys("/home/makarand/workspace/Practise/src/main/resources/pic.jpg"); // please provide absolute path of the file to upload.

    }
```

**Comment by andreastt** [comment]
Can you include the verbose log output from `-vv`? File uploads work in the Marionette tests, but it’s possible Selenium has slightly different protocol semantics that differs from the WebDriver specification.

**Comment by MakarandDhuri** [comment]
Hi @andreastt ,

Please find logs  below 

```
1470742798119	Marionette	INFO	Listening on port 50060
1470742800396	Marionette	INFO	startBrowser f161adb0-bde2-4aa6-85c0-65e8235873ca
1470742800412	Marionette	INFO	sendAsync f161adb0-bde2-4aa6-85c0-65e8235873ca

(process:13924): GLib-CRITICAL **: g_path_get_basename: assertion 'file_name != NULL' failed
1470742800871	Marionette	INFO	sendAsync f161adb0-bde2-4aa6-85c0-65e8235873ca
1470742801098	Marionette	INFO	sendAsync f161adb0-bde2-4aa6-85c0-65e8235873ca
1470742808472	Marionette	INFO	sendAsync f161adb0-bde2-4aa6-85c0-65e8235873ca
1470742808486	Marionette	INFO	sendAsync f161adb0-bde2-4aa6-85c0-65e8235873ca
1470742808501	Marionette	INFO	sendAsync f161adb0-bde2-4aa6-85c0-65e8235873ca
1470742808510	Marionette	INFO	sendAsync f161adb0-bde2-4aa6-85c0-65e8235873ca
1470742808518	Marionette	INFO	sendAsync f161adb0-bde2-4aa6-85c0-65e8235873ca
```

**Comment by andreastt** [comment]
Unfortunately you only included the INFO-level messages. To be able to diagnose this correctly you need to include the TRACE-level messages. You can include those by using the `-vv` or `--log TRACE` arguments to geckodriver.

**Comment by shane-tomlinson** [comment]
@andreastt, @jgraham - Here's my log with the same problem:

```
15:38:24.808 INFO - Done: [find element: By.cssSelector: #file]
15:38:24.811 INFO - Executing: [click: 14 [[FirefoxDriver: Firefox on MAC (212df310-477e-524c-89f1-ec1d42e61e0a)] -> css selector: #file]])
1470926304812	Marionette	TRACE	conn0 -> [0,79,"clickElement",{"id":"3b31c1b0-dc8b-9b40-8b87-cd807d093102"}]
1470926304813	Marionette	INFO	sendAsync 212df310-477e-524c-89f1-ec1d42e61e0a
1470926304829	Marionette	TRACE	conn0 <- [1,79,null,{}]
15:38:25.146 INFO - Done: [click: 14 [[FirefoxDriver: Firefox on MAC (212df310-477e-524c-89f1-ec1d42e61e0a)] -> css selector: #file]]
15:38:25.151 INFO - Executing: [find element: By.cssSelector: #imageLoader])
1470926305470	Marionette	TRACE	conn0 -> [0,80,"findElement",{"using":"css selector","value":"#imageLoader"}]
1470926305471	Marionette	INFO	sendAsync 212df310-477e-524c-89f1-ec1d42e61e0a
1470926305473	Marionette	TRACE	conn0 <- [1,80,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"4a92d67a-7fd5-9246-ae59-126aeef4f71f","ELEMENT":"4a92d67a-7fd5-9246-ae59-126aeef4f71f"}}]
15:38:25.478 INFO - Done: [find element: By.cssSelector: #imageLoader]
15:38:25.487 INFO - Executing: org.openqa.selenium.remote.server.handler.UploadFile@1e51bf25)
15:38:25.488 INFO - Done: org.openqa.selenium.remote.server.handler.UploadFile@1e51bf25
15:38:25.491 INFO - Executing: [send keys: 13 [[FirefoxDriver: Firefox on MAC (212df310-477e-524c-89f1-ec1d42e61e0a)] -> css selector: #imageLoader], [/var/folders/x6/zd3bzsyd08v932ymg7t283rc0000gn/T/299935fc-8b32-4554-a52d-71473f95383b/upload5501428791666416436file/apple-touch-icon-152x152.png]])
1470926305493	Marionette	TRACE	conn0 -> [0,81,"sendKeysToElement",{"id":"4a92d67a-7fd5-9246-ae59-126aeef4f71f","value":["/","v","a","r","/","f","o","l","d","e","r","s","/","x","6","/","z","d","3","b","z","s","y","d","0","8","v","9","3","2","y","m","g","7","t","2","8","3","r","c","0","0","0","0","g","n","/","T","/","2","9","9","9","3","5","f","c","-","8","b","3","2","-","4","5","5","4","-","a","5","2","d","-","7","1","4","7","3","f","9","5","3","8","3","b","/","u","p","l","o","a","d","5","5","0","1","4","2","8","7","9","1","6","6","6","4","1","6","4","3","6","f","i","l","e","/","a","p","p","l","e","-","t","o","u","c","h","-","i","c","o","n","-","1","5","2","x","1","5","2",".","p","n","g"]}]
1470926305494	Marionette	INFO	sendAsync 212df310-477e-524c-89f1-ec1d42e61e0a
1470926305494	Marionette	INFO	sendAsync 212df310-477e-524c-89f1-ec1d42e61e0a
1470926305495	Marionette	TRACE	conn0 <- [1,81,null,{}]
15:38:25.496 INFO - Done: [send keys: 13 [[FirefoxDriver: Firefox on MAC (212df310-477e-524c-89f1-ec1d42e61e0a)] -> css selector: #imageLoader], [/var/folders/x6/zd3bzsyd08v932ymg7t283rc0000gn/T/299935fc-8b32-4554-a52d-71473f95383b/upload5501428791666416436file/apple-touch-icon-152x152.png]]
15:38:25.499 INFO - Executing: [find element: By.cssSelector: .cropper])
1470926305500	Marionette	TRACE	conn0 -> [0,82,"findElement",{"using":"css selector","value":".cropper"}]
1470926305501	Marionette	INFO	sendAsync 212df310-477e-524c-89f1-ec1d42e61e0a
1470926333507	Marionette	TRACE	conn0 <- [1,82,{"error":"no such element","message":"Unable to locate element: .cropper","stacktrace":""},null]
15:38:53.969 WARN - Exception thrown
org.openqa.selenium.NoSuchElementException: Unable to locate element: .cropper (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 28.01 seconds
```

**Comment by acikojevic** [comment]
Hi, facing the same issue here, sendKeys upload is not working on geckodriver, but chromedriver and iedriver are working normally. I'm using RemoteWebDriver.
OS: WIN8_1
Geckodriver: v0.10.0-win64
Selenium server: selenium-server-standalone-3.0.0-beta3
Browser: FF 48.0.2 32bit, tried with 48.0.2 64 bit, 47.0.1 32/64 bit, always getting same results

These are some logs, it seems fine, but upload is not working, nothing happens.

09:20:13.629 INFO - Executing: [send keys: 6 [[[[FirefoxDriver: Firefox on XP (80421bbb-0c3a-45f4-be3a-398e22c98295)] -> css selector: div#createCampaign div.ib-fileupload]] -> css selector: input.upload[name=file]], [C, :, \, S, e, l, e, n, i, u, m, \, T, e, s, t, D, a, t, a, \, C, U, P, \, T, C, _, C, r, e, a, t, e, C, a, m, p, a, i, g, n, _, M, T, _, F, i, l, e, 2, S, M, S, _, C, o, u, n, t, r, y, P, r, e, f, i, x, ., x, l, s, x]])
09:20:13.629 DEBUG - Executing: sendKeysToElement [80421bbb-0c3a-45f4-be3a-398e22c98295, sendKeysToElement {id=9dcbe611-e4af-4a9b-bd6f-971c0562ec05, value=[Ljava.lang.CharSequence;@52fbba6}]
09:20:13.631 DEBUG - CookieSpec selected: default
09:20:13.631 DEBUG - Auth cache not set in the context
09:20:13.634 DEBUG - Connection request: [route: {}->http://localhost:2449][total kept alive: 0; route allocated: 0 of 2000; total allocated: 0 of 2000]
09:20:13.635 DEBUG - Connection leased: [id: 49][route: {}->http://localhost:2449][total kept alive: 0; route allocated: 1 of 2000; total allocated: 1 of 2000]
09:20:13.635 DEBUG - Opening connection {}->http://localhost:2449
09:20:13.636 DEBUG - Connecting to localhost/127.0.0.1:2449
09:20:13.637 DEBUG - Connection established 127.0.0.1:56420<->127.0.0.1:2449
09:20:13.637 DEBUG - http-outgoing-49: set socket timeout to 10800000
09:20:13.637 DEBUG - Executing request POST /session/80421bbb-0c3a-45f4-be3a-398e22c98295/element/9dcbe611-e4af-4a9b-bd6f-971c0562ec05/value HTTP/1.1
09:20:13.638 DEBUG - Target auth state: UNCHALLENGED
09:20:13.638 DEBUG - Proxy auth state: UNCHALLENGED
09:20:13.638 DEBUG - http-outgoing-49 >> POST /session/80421bbb-0c3a-45f4-be3a-398e22c98295/element/9dcbe611-e4af-4a9b-bd6f-971c0562ec05/value HTTP/1.1
09:20:13.639 DEBUG - http-outgoing-49 >> Content-Type: application/json; charset=utf-8
09:20:13.639 DEBUG - http-outgoing-49 >> Content-Length: 351
09:20:13.639 DEBUG - http-outgoing-49 >> Host: localhost:2449
09:20:13.639 DEBUG - http-outgoing-49 >> Connection: Keep-Alive
09:20:13.639 DEBUG - http-outgoing-49 >> User-Agent: Apache-HttpClient/4.5.2 (Java/1.8.0_101)
09:20:13.640 DEBUG - http-outgoing-49 >> Accept-Encoding: gzip,deflate
09:20:13.640 DEBUG - http-outgoing-49 >> "POST /session/80421bbb-0c3a-45f4-be3a-398e22c98295/element/9dcbe611-e4af-4a9b-bd6f-971c0562ec05/value HTTP/1.1[\r][\n]"
09:20:13.640 DEBUG - http-outgoing-49 >> "Content-Type: application/json; charset=utf-8[\r][\n]"
09:20:13.640 DEBUG - http-outgoing-49 >> "Content-Length: 351[\r][\n]"
09:20:13.641 DEBUG - http-outgoing-49 >> "Host: localhost:2449[\r][\n]"
09:20:13.641 DEBUG - http-outgoing-49 >> "Connection: Keep-Alive[\r][\n]"
09:20:13.641 DEBUG - http-outgoing-49 >> "User-Agent: Apache-HttpClient/4.5.2 (Java/1.8.0_101)[\r][\n]"
09:20:13.641 DEBUG - http-outgoing-49 >> "Accept-Encoding: gzip,deflate[\r][\n]"
09:20:13.641 DEBUG - http-outgoing-49 >> "[\r][\n]"
09:20:13.642 DEBUG - http-outgoing-49 >> "{"id":"9dcbe611-e4af-4a9b-bd6f-971c0562ec05","value":["C",":","\\","S","e","l","e","n","i","u","m","\\","T","e","s","t","D","a","t","a","\\","C","U","P","\\","T","C","_","C","r","e","a","t","e","C","a","m","p","a","i","g","n","_","M","T","_","F","i","l","e","2","S","M","S","_","C","o","u","n","t","r","y","P","r","e","f","i","x",".","x","l","s","x"]}"
1473405613658   Marionette      INFO    sendAsync 80421bbb-0c3a-45f4-be3a-398e22c98295
1473405613700   Marionette      INFO    sendAsync 80421bbb-0c3a-45f4-be3a-398e22c98295
09:20:13.722 DEBUG - http-outgoing-49 << "HTTP/1.1 200 OK[\r][\n]"
09:20:13.723 DEBUG - http-outgoing-49 << "Content-Length: 2[\r][\n]"
09:20:13.723 DEBUG - http-outgoing-49 << "Content-Type: application/json[\r][\n]"
09:20:13.723 DEBUG - http-outgoing-49 << "Date: Fri, 09 Sep 2016 07:20:13 GMT[\r][\n]"
09:20:13.724 DEBUG - http-outgoing-49 << "Connection: close[\r][\n]"
09:20:13.724 DEBUG - http-outgoing-49 << "[\r][\n]"
09:20:13.724 DEBUG - http-outgoing-49 << "{}"
09:20:13.725 DEBUG - http-outgoing-49 << HTTP/1.1 200 OK
09:20:13.725 DEBUG - http-outgoing-49 << Content-Length: 2
09:20:13.725 DEBUG - http-outgoing-49 << Content-Type: application/json
09:20:13.725 DEBUG - http-outgoing-49 << Date: Fri, 09 Sep 2016 07:20:13 GMT
09:20:13.725 DEBUG - http-outgoing-49 << Connection: close
09:20:13.726 DEBUG - http-outgoing-49: Close connection
09:20:13.727 DEBUG - Connection discarded
09:20:13.727 DEBUG - Connection released: [id: 49][route: {}->http://localhost:2449][total kept alive: 0; route allocated: 0 of 2000; total allocated: 0 of 2000]
09:20:13.728 DEBUG - Executed: [80421bbb-0c3a-45f4-be3a-398e22c98295, sendKeysToElement {id=9dcbe611-e4af-4a9b-bd6f-971c0562ec05, value=[Ljava.lang.CharSequence;@52fbba6}]
09:20:13.728 INFO - Done: [send keys: 6 [[[[FirefoxDriver: Firefox on XP (80421bbb-0c3a-45f4-be3a-398e22c98295)] -> css selector: div#createCampaign div.ib-fileupload]] -> css selector: input.upload[name=file]], [C, :, \, S, e, l, e, n, i, u, m, \, T, e, s, t, D, a, t, a, \, C, U, P, \, T, C, _, C, r, e, a, t, e, C, a, m, p, a, i, g, n, _, M, T, _, F, i, l, e, 2, S, M, S, _, C, o, u, n, t, r, y, P, r, e, f, i, x, ., x, l, s, x]]
09:20:13.729 DEBUG - Finished: POST /session/d74ba688-f277-4b70-9f14-89317065248d/element/6/value

**Comment by andreastt** [comment]
I just confirmed with [wdclient](https://github.com/w3c/wdclient) and geckodriver against a Nightly build that I’m able to set an `<input type=file>`’s value by sending keys. This proves that the problem isn’t with geckodriver.

**Comment by andreastt** [comment]
For what it’s worth I also downloaded Firefox 48 and ran the same test there.

**Comment by shane-tomlinson** [comment]
> I just confirmed with wdclient and geckodriver against a Nightly build that I’m able to set an <input type=file>’s value by sending keys. This proves that the problem isn’t with geckodriver.

Is this after opening the file selection modal, or just sending the keys to the input directly w/o opening the modal? Can I see a copy of your test to try and mirror as closely as possible in our own tests to see if it's just a problem of not sending the same commands?

**Comment by acikojevic** [comment]
@andreastt @shane-tomlinson , tnx for checking this out. Just to let you know I can't get any file selection modal opened, nothing happens after sendKeys is executed. Today I tried using FirefoxDriver explicitly, without any capabilities or preferences, and it's still not working just like when RemoteWebDriver is used. On selenium 2.53.1 everything works fine (no geckodriver is used), just like on IE/chrome drivers. Tried with different versions and options so I'm out of ideas how to solve this if the problem is on my side. :)

**Comment by acikojevic** [comment]
Problem is solved when using selenium-server-standalone-3.0.0**-beta4**! Firefox v49.0.1.

**Comment by shane-tomlinson** [comment]
> Problem is solved with Firefox v49.0.1.
> Tried on selenium-3.0.0-beta3 and selenium-3.0.0-beta4.

This is the best news I've read all morning!

**Comment by nivithaCisco** [comment]
i am facing the problem still

**Comment by andreastt** [comment]
@nivithaCisco It is impossible to help you with that little information… ensure you are using the latest version of geckodriver and [Firefox Nightly](http://nightly.mozilla.org/).

**Comment by AndyUM357** [comment]
I'm seeing the same issue in Firefox v50/GeckoDriverv11.1 with WebDriver v3.0.1 using the RemoteWebDriver.  When compared to Chrome (which works), something I noticed is the splitting of the send-keys "value" into an array of characters.  Chrome does not do that.

You can see the same thing with the debug post from @acikojevic above:
`09:20:13.642 DEBUG - http-outgoing-49 >> "{"id":"9dcbe611-e4af-4a9b-bd6f-971c0562ec05","value":["C",":","\\","S","e","l","e","n","i","u","m","\\","T","e","s","t","D","a","t","a","\\","C","U","P","\\","T","C","_","C","r","e","a","t","e","C","a","m","p","a","i","g","n","_","M","T","_","F","i","l","e","2","S","M","S","_","C","o","u","n","t","r","y","P","r","e","f","i","x",".","x","l","s","x"]}"`

Should the transmission be a "value" array of one solid (backslash-delimited) string?  Like this:
`"{"id":"9dcbe611-e4af-4a9b-bd6f-971c0562ec05","value":["C:\\Selenium\\TestData\\CUP\\TC_CreateCampaign_MT_File2SMS_CountryPrefix.xslx"]}"`

Thoughts?

**Comment by andreastt** [comment]
Yes. Chrome is currently doing the wrong thing according to the specification.

However, the specification will change to send one string once I have had time to update it. When that is done, we need to do a compatibility dance with webdriver-rust, geckodriver, the Marionette Python client, the Firefox UI tests, the Marionette tests, and Selenium.

**Comment by AndyUM357** [comment]
What also seems suspicious is prior to the send-keys there should be an UploadFile call which uploads the file to the node so it can access it locally.  I see the UploadFile as successful in the logs, but I cannot find the file on the node.  I can do these steps with Chrome as a comparison and find the file (for me, it stuffs it in a User\AppData\Local\Temp area).  The logs claim the file upload transmission worked, but it may not have.

**Comment by stephaneblais** [comment]
I'm using selenium 3.0.1 geckodriver 0.11.1 and FF 50.1.0 and the upload works like a charm but for some reason if I try to upload 2 files in a row it always upload the first one twice mmmm

File file = new File(rP.getInfoFromFile("uploadPhoneBillPath"));
upbp.selectAttachmentFile(file.getAbsolutePath());
file = new File(rP.getInfoFromFile("uploadPhoneBillPath2"));
upbp.selectAttachmentFile(file.getAbsolutePath());

**Comment by andreastt** [comment]
@stephaneblais That is odd, but it seems like a disparate issue. Do you mind opening a new issue about that?

**Comment by andreastt** [comment]
Also, since @stephaneblais confirmed it is working with the most recent Firefox, geckodriver, and Selenium versions I’m going to close this issue.

Like I said earlier, geckodriver and Marionette has supported file uploads for a very long time, but Selenium and various client bindings have been sending the wrong input according to the WebDriver specification.

**Comment by tomprakash** [comment]
I am facing the same problem with below details when trying to upload file using AutoIT.
OS: Windows 7
Browser: Firefox 59.0.2(64-bit)
selenium version: 3.11.0

Apr 11, 2018 11:16:30 PM org.openqa.selenium.remote.DesiredCapabilities firefox
INFO: Using `new FirefoxOptions()` is preferred to `DesiredCapabilities.firefox()`
1523468791313	geckodriver	INFO	geckodriver 0.20.0
1523468791328	geckodriver	INFO	Listening on 127.0.0.1:11567
1523468792117	mozrunner::runner	INFO	Running command: "C:\\Program Files\\Mozilla Firefox\\firefox.exe" "-marionette" "-profile" "C:\\Users\\Tom\\AppData\\Local\\Temp\\rust_mozprofile.ErGbGT1wIMN2"
1523468792946	Marionette	INFO	Enabled via --marionette
1523468796232	Marionette	INFO	Listening on port 51778
1523468797402	Marionette	WARN	TLS certificate errors will be ignored for this session
Apr 11, 2018 11:16:37 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
Unable to read VR Path Registry from C:\Users\Tom\AppData\Local\openvr\openvrpaths.vrpath
[Child 8720, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
Exception in thread "main" org.openqa.selenium.InvalidArgumentException: Cannot click <input type=file> elements
Build info: version: '3.11.0', revision: 'e59cfb3', time: '2018-03-11T20:33:08.638Z'
System info: host: 'TOM-PC', ip: '10.0.0.2', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '1.8.0_161'
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Capabilities {acceptInsecureCerts: true, browserName: firefox, browserVersion: 59.0.2, javascriptEnabled: true, moz:accessibilityChecks: false, moz:headless: false, moz:processID: 8360, moz:profile: C:\Users\Tom\AppData\Local\..., moz:useNonSpecCompliantPointerOrigin: false, moz:webdriverClick: true, pageLoadStrategy: normal, platform: XP, platformName: XP, platformVersion: 6.1, rotatable: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}}
Session ID: c298bac9-f62c-4963-b0e9-8acbd2500b08
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Unknown Source)
	at org.openqa.selenium.remote.http.W3CHttpResponseCodec.createException(W3CHttpResponseCodec.java:187)
	at org.openqa.selenium.remote.http.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:122)
	at org.openqa.selenium.remote.http.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:49)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:158)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:83)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:545)
	at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:279)
	at org.openqa.selenium.remote.RemoteWebElement.click(RemoteWebElement.java:83)
	at automationframework.FileUpload.main(FileUpload.java:35)

**Comment by andreastt** [comment]
Please don’t hijack old, resolved issues.

You shouldn’t use AutoIT to upload files, but instead use the Element Send Keys command on the `<input type=file>` element.
